substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeBootImage.java

```diff
@@ -952,4 +973,299 @@ protected NativeTextSectionImpl(RelocatableBuffer relocatableBuffer, ObjectFile
+     * that allows type, code and heap data info to be passed to
+     * an ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugInfoProvider implements DebugInfoProvider {
+        private final NativeImageCodeCache codeCache;
+        private final NativeImageHeap heap;
+        private final Iterator<Map.Entry<HostedMethod, CompilationResult>> codeCacheIterator;
+        private final Iterator<Map.Entry<Object, ObjectInfo>> heapIterator;
+
+        NativeImageDebugInfoProvider(NativeImageCodeCache codeCache, NativeImageHeap heap) {
+            super();
+            this.codeCache = codeCache;
+            this.heap = heap;
+            this.codeCacheIterator = codeCache.compilations.entrySet().iterator();
+            this.heapIterator = heap.objects.entrySet().iterator();
+        }
+
+        @Override
+        public DebugTypeInfoProvider typeInfoProvider() {
+            return () -> new Iterator<DebugTypeInfo>() {
+                @Override
+                public boolean hasNext() {
+                    return false;
+                }
+
+                @Override
+                public DebugTypeInfo next() {
+                    return null;
+                }
+            };
+        }
+
+        @Override
+        public DebugCodeInfoProvider codeInfoProvider() {
+            return () -> new Iterator<DebugCodeInfo>() {
+                @Override
+                public boolean hasNext() {
+                    return codeCacheIterator.hasNext();
+                }
+
+                @Override
+                public DebugCodeInfo next() {
+                    Map.Entry<HostedMethod, CompilationResult> entry = codeCacheIterator.next();
+                    return new NativeImageDebugCodeInfo(entry.getKey(), entry.getValue());
+                }
+            };
+        }
+
+        @Override
+        public DebugDataInfoProvider dataInfoProvider() {
+            return () -> new Iterator<DebugDataInfo>() {
+                @Override
+                public boolean hasNext() {
+                    return false;
+                }
+
+                @Override
+                public DebugDataInfo next() {
+                    return null;
+                }
+            };
+        }
+    }
+
+    /**
+     * implementation of the DebugCodeInfo API interface
+     * that allows code info to be passed to an ObjectFile
+     * when generation of debug info is enabled.
+     */
+    private class NativeImageDebugCodeInfo implements DebugCodeInfo {
+        private final HostedMethod method;
+        private final CompilationResult compilation;
+
+        NativeImageDebugCodeInfo(HostedMethod method, CompilationResult compilation) {
+            this.method = method;
+            this.compilation = compilation;
+        }
+
+        @Override
+        public String fileName() {
+            HostedType declaringClass = method.getDeclaringClass();
```

Comment:
In Java 11 the module base-name is also part of the full file path with the src.zip. e.g.
`src.zip` contains `/java.base/java/util/Set.class`

Reply:
Well, I finally managed to rebuild using jdk11 (I had to nuke all the mxbuild dirs to get it to stop using cached jdk8 code) and found out some interesting things about pathe that probably need to be discussed and decided upon.

I tweaked the path generation to add the result of ModuleSupport.getModuleName(javaClass) to the head of the path expression. So, for jdk11 runtime classes that will add e.g. "java.base" to the list of Path elements whereas for jdk8 runtime classes it will add nothing since  the module name is returned as "". This means that unzipped jdk sources are now being located correctly on both jdk8 and jdk11 assuming the root dir for the unzipped src.zip is added to the source path.

I prefixed the module name to the file path for all classes that belong to a module on the assumption that this action would only apply for JDK runtime code. However, this assumption is not strictly correct as some Graal classes reside in JDK runtime modules.

I noticed that, for example, class AllocationSnippets in package org/graalvm/compiler/replacements turns out to belong to module jdk.internal.vm.compiler. The source for this class actually sits in the Graal tree under root compiler/src/org.graalvm.compiler.replacements/src. Another example is class PrefetchAllocateNode which also belongs to module jdk.internal.vm.compiler. The source for this class also sits in the Graal tree under root compiler/src/org.graalvm.compiler.nodes/src.

I also noticed that various of the Graal class sources are bundled into the src.zip for the jdk11 release I am using (labsjdk-ce-11.0.6+9-jvmci-20.0-b02-linux-amd64) and indeed are bundled into stock jdk11u releases. The source for PrefetchAllocateNode is present under root jdk.internal.vm.compiler/org.graalvm.compiler.nodes/src. However, the source for AllocationSnippets is not present. It looks as if these are present by because they are present in the OpenJDK repo (which may explain why AllocationSnippets.java is missing).

So, in order to find the correct versions of these sources from the Graal tree rather than from the OpenJDK src.zip it appears I will need to special case classes that are in certain well-known org.graal and com.oracle packages and not prefix the module to the file path for those cases. For example, the file path for PrefetchAllocateNode will need to be org/graalvm/compiler/nodes rather than jdk.internal.vm.compiler/org/graalvm/compiler/nodes.

In order to minimize the number of source roots needed to locate Graal sources it might also be helpful to insert the required overarching src root into the file path. For example, PrefetchAllocateNode might be better associated with file path org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/. That would  mean that only  compiler/src, substratevm/src etc would need to be supplied as source root dirs. 

The difficutly with this is that it is not always clear what the oevrarching top level directory needs to be. For example, class CodeInfo.java sits under dir com.oracle.svm.core/src but resides in subdir com/oracle/svm/core/code because its is in a subpackage com.oracle.svm.core.code. Given any Graal class the overarching root dir always comprises some initial segment of the package name but it is not always clear which elements to include. This is slightly complicated by the need to recognize that classes in package com.oracle.svm.core.code map to root dir com.oracle.svm.core/src but that classes in package com.oracle.svm.core.graal map to root dir com.oracle.svm.core.graal/src

I did actually implement a variant of the filepath computation some while back which used a hard-wired list of known root packages. I could easily do that again. However, the problem is keeping the list of source roots up to date with any changes. Do you have any other suggestions as to how to do this? Input from other Graal community devs woudl be very welcome.

Reply:
> This means that unzipped jdk sources are now being located correctly on both jdk8 and jdk11 assuming the root dir for the unzipped src.zip is added to the source path.

Unzipping should happen transparently. Just use Zip File System Provider instead of requiring the user to unzip src.zip. See https://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html - For examples in our code base where we make use of Zip File System Provider see e.g. com.oracle.svm.driver.NativeImage#processClasspathNativeImageMetaInf(java.nio.file.Path, com.oracle.svm.driver.NativeImage.NativeImageMetaInfResourceProcessor)

Reply:
> Another example is class PrefetchAllocateNode which also belongs to module jdk.internal.vm.compiler. The source for this class also sits in the Graal tree under root compiler/src/org.graalvm.compiler.nodes/src.

As you noticed even vanilla JDK11 releases contain Graal. For debuginfo generation this version of Graal should be ignored (it exists so that vanilla JDK can provide Graal as tier-4 JIT compiler when using `-XX:+UnlockExperimentalVMOptions` `-XX:+UseJVMCICompiler`). When we build native images we always use the version of Graal that is shipped as part of a GraalVM release. Thus any Graal related source files that exist in src.zip are irrelevant for us.

Reply:
> However, the problem is keeping the list of source roots up to date with any changes.

Sure you could diff each file do decide if you need to update a preexisting target file but it is probably easier to simply copy over based on timestamps (if src or zip file that contains src is newer -> update)

Reply:
> source roots needed to locate Graal sources

A good heuristic for the roots is to inspect the jar files that are on the image classpath and check if you find a corresponding sibling file. E.g. for `<graalvm>/lib/svm/builder/svm.jar` you probe for `<graalvm>/lib/svm/builder/svm.src.zip`. If the classpath entry is a directory `<dir>` instead you probe for `<dir>/src` and `<dir>/src_gen` as roots. For each source file lookup you check against these roots for a match. (`src.zip` from jdk is a special case and should always be in the list of roots (obviously as last entry))

Reply:
> Unzipping should happen transparently. Just use Zip File System Provider instead of requiring the user to unzip src.zip. See https://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html - For examples in our code base where we make use of Zip File System Provider see e.g. com.oracle.svm.driver.NativeImage#processClasspathNativeImageMetaInf(java.nio.file.Path, com.oracle.svm.driver.NativeImage.NativeImageMetaInfResourceProcessor)

I'm not clear what you are suggesting here. I understand how to use a zip file provider from Java. However, the problem I am looking at is how to allow gdb to locate a source file for a JDK class when it is running. gdb needs to be pointed at a root directory under which it can locate the sources. It is not going to unzip any files to do that.

I believe the model you are proposing for resolving that requirement is that the native image generator should attempt to locate all the relevant JDK, Graal (and, perhaps, application) source files at image build time, bundling them into a jar/zip file provided as a secondary deliverable. If that jar/zip deliverable is correctly structured then it could be unzipped into a single root source directory from which all file paths and name nstalled into the debuginfo could be successfully resolved to the required source file at debug time. Is that a fair summary?

> A good heuristic for the roots is to inspect the jar files that are on the image classpath and check if you find a corresponding sibling file. E.g. for <graalvm>/lib/svm/builder/svm.jar you probe for <graalvm>/lib/svm/builder/svm.src.zip. If the classpath entry is a directory <dir> instead you probe for <dir>/src and <dir>/src_gen as roots. For each source file lookup you check against these roots for a match. (src.zip from jdk is a special case and should always be in the list of roots (obviously as last entry))

Yes, I am sure I can implement a lookup model like that.

> As you noticed even vanilla JDK11 releases contain Graal.  . . . Thus any Graal related source files that exist in src.zip are irrelevant for us.

Yes, I will ignore those files.

Reply:
> I'm not clear what you are suggesting here.

I'm just saying the mechanism should be able to process zip-files, jar-files & directories alike as inputs (roots) and copy files referenced in debuginfo (`--x-->`) from those roots into one single output directory (preferably named `sources`) (the one that gdb will then use for file retrival):
```
[... foo/prjx/src]
[... foo/lib/gson.sources.jar]
[...]                                           --x-->     [... foo/sources]
[<graalvm>/lib/svm/builder/svm.src.zip]
[<graalvm>/src.zip]
```

Reply:
Ok, that sounds fine. I'll work on adding that.

Reply:
n.b. I just updated the issue to record this conversation with a more precise 'spec'

Reply:
This has now been implemented -- see my other comment